src/sundials/sundials_adaptcontroller.c

```diff
@@ -0,0 +1,173 @@
+
+void SUNAdaptController_FreeEmpty(SUNAdaptController C)
+{
+  if (C == NULL)  return;
+
+  /* free non-NULL ops structure */
+  if (C->ops)  free(C->ops);
+  C->ops = NULL;
+
+  /* free overall SUNAdaptController object and return */
+  free(C);
+  return;
+}
+
+
+/* -----------------------------------------------------------------
+ * Required functions in the 'ops' structure for non-NULL controller
+ * ----------------------------------------------------------------- */
+
+SUNAdaptController_Type SUNAdaptController_GetType(SUNAdaptController C)
+{
+  if (C == NULL) { return SUN_ADAPTCONTROLLER_NONE; }
+  return(C->ops->gettype(C));
+}
+
+/* -----------------------------------------------------------------
+ * Optional functions in the 'ops' structure
+ * ----------------------------------------------------------------- */
+
+int SUNAdaptController_Destroy(SUNAdaptController C)
+{
+  if (C == NULL) return(SUNADAPTCONTROLLER_SUCCESS);
+
+  /* if the destroy operation exists use it */
+  if (C->ops)
+    if (C->ops->destroy) { return(C->ops->destroy(C)); }
+
+  /* if we reach this point, either ops == NULL or destroy == NULL,
+     try to cleanup by freeing the content, ops, and matrix */
+  if (C->content) { free(C->content); C->content = NULL; }
+  if (C->ops) { free(C->ops); C->ops = NULL; }
+  free(C); C = NULL;
+
+  return(SUNADAPTCONTROLLER_SUCCESS);
+}
+
+int SUNAdaptController_EstimateStep(SUNAdaptController C, sunrealtype h, int p,
+                                    sunrealtype dsm, sunrealtype* hnew)
+{
+  int ier = 0;
+  *hnew = h;   /* initialize output with identity */
+  if (C == NULL) { return ier; }
+  if (C->ops->estimatestep)
+  {
+    ier = C->ops->estimatestep(C, h, p, dsm, hnew);
+  }
+  return(ier);
+}
+
+
+int SUNAdaptController_Reset(SUNAdaptController C)
+{
+  int ier = 0;
+  if (C == NULL) { return ier; }
+  if (C->ops->reset) { ier = C->ops->reset(C); }
+  return(ier);
+}
+
+int SUNAdaptController_SetDefaults(SUNAdaptController C)
+{
+  int ier = 0;
+  if (C->ops->setdefaults) { ier = C->ops->setdefaults(C); }
+  return(ier);
+}
+
+int SUNAdaptController_Write(SUNAdaptController C, FILE* fptr)
+{
+  int ier = 0;
+  if (C == NULL) { return ier; }
+  if (C->ops->write) { ier = C->ops->write(C, fptr); }
+  return(ier);
+}
+
+int SUNAdaptController_SetErrorBias(SUNAdaptController C, sunrealtype bias)
+{
+  int ier = 0;
+  if (C == NULL) { return ier; }
+  if (C->ops->seterrorbias) { ier = C->ops->seterrorbias(C, bias); }
+  return(ier);
+}
+
+int SUNAdaptController_UpdateH(SUNAdaptController C, sunrealtype h, sunrealtype dsm)
+{
+  int ier = 0;
+  if (C == NULL) { return ier; }
+  if (C->ops->updateh) { ier = C->ops->updateh(C, h, dsm); }
+  return(ier);
+}
+
+int SUNAdaptController_Space(SUNAdaptController C, long int *lenrw, long int *leniw)
+{
+  int ier = 0;
+  *lenrw = 0;   /* initialize outputs with identity */
+  *leniw = 0;
+  if (C == NULL) { return ier; }
```

Comment:
```suggestion
  if (C == NULL || lenrw == NULL || leniw == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
  *lenrw = 0;   /* initialize outputs with identity */
  *leniw = 0;
```

Reply:
I don't see why these checks on `lenrw` and `leniw` need to be duplicated everywhere. If it is here then there is no reason to do it redundantly in each implementation (as requested).

Reply:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.